File: python/day1/main.py
import itertools
import functools
import re
import math

from copy import deepcopy
from collections import defaultdict

INPUT_FILE = 'input.txt'

def read_file(filename, func=None):
    result = []
    with open(filename, 'r') as f:
        for line in f:
            result.append(func(line.strip()) if func else line.strip())
    return result

def get_elves():
    lines = read_file(INPUT_FILE)
    elves = []
    current = 0
    for l in lines:
        if not l:
            elves.append(current)
            current = 0
            continue
        current += int(l)
    return sorted(elves, reverse=True)

def main2():
    elves = get_elves()
    print(sum(elves[:3]))

def main1():
    elves = get_elves()
    print(elves[0])

def inline_solution():
    lines = open(INPUT_FILE, 'r').read().split('\n\n')
    elves = sorted([sum(map(int, elf.split())) for elf in lines], reverse=True)
    print(elves[0])
    print(sum(elves[:3]))

if __name__ == "__main__":
    main1()
    main2()

File: rust/day1/src/main.rs
use std::fs::File;
use std::io::Read;

fn main() {
    part1();
    part2();
}

fn compute() -> Vec<i32> {
    let mut file = File::open("input.txt").unwrap();
    let mut s = String::new();
    file.read_to_string(&mut s).unwrap();
    let vec: Vec<&str> = s.split("\n\n").collect();
    let mut elves: Vec<i32> = vec.iter()
        .map(|item| item.split("\n")
             .filter(|x| !x.is_empty())
             .map(|x| x.parse::<i32>().unwrap()).sum::<i32>()
        )
        .collect();
    elves.sort_by(|a, b| b.cmp(a));
    elves
}

fn part1() {
    let elves = compute();
    println!("{:?}", elves[0]);
}

fn part2() {
    let elves = compute();
    println!("{:?}", &elves[0..3].iter().sum::<i32>());
}
